### PR TITLE
samples: doc: Added README for POSIX samples

### PR DESCRIPTION
**Pull request by dougwfoster**
Added README.rst for eventfd and uname POSIX samples.
Helps fix #27805

### CODE REVIEW

**Review by cfriedt — DISMISSED** [review]
LGTM - a great follow up commit or PR would be to list POSIX samples in the os services / portability / POSIX doc area.

Would be happy to re-approve if you wish to add that here, @dougwfoster ?

File: samples/posix/eventfd/README.rst [hunk 1]
```diff
@@ -0,0 +1,46 @@
+.. _posix-eventfd-sample:
+
+POSIX eventfd()
+########################################################
```
Comment (ycsin) [inline samples/posix/eventfd/README.rst:4]: IIRC, their length should match?

```suggestion
POSIX eventfd()
###############
```

File: samples/posix/uname/README.rst [hunk 2]
```diff
@@ -0,0 +1,56 @@
+.. _posix-uname-sample:
+
+POSIX uname()
+########################################################
```
Comment (ycsin) [inline samples/posix/uname/README.rst:4]:
```suggestion
POSIX uname()
#############
```

File: samples/posix/uname/README.rst [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+Overview
+********
+
+In this sample application, the POSIX uname() function
```
Comment (kartben) [inline samples/posix/uname/README.rst:9]: adding an hyperlink to uname()'s doc would be welcome
https://pubs.opengroup.org/onlinepubs/9699919799/functions/uname.html

File: samples/posix/eventfd/README.rst [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+Overview
+********
+
+This sample application demonstrates using the POSIX eventfd()
```
Comment (kartben) [inline samples/posix/eventfd/README.rst:9]: adding a reference to eventfd() doc would be welcome here, but not sure why there's none on opengroup webpage? @cfriedt might know

File: samples/posix/eventfd/README.rst [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+function to create a file descriptor, which can be used for event notification.
+The returned file descriptor is used with write/read calls and write/read values
+are output to the console.
+
+
```
Comment (kartben) [inline samples/posix/eventfd/README.rst:14]: nit: remove extra line

```suggestion
```

File: samples/posix/uname/README.rst [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+is used to acquire system information and it is output to
+the console. Additionally, uname is added as a shell command and system
+information is displayed according to the option(s) provided for the command.
+
+
```
Comment (kartben) [inline samples/posix/uname/README.rst:14]: nit: remove

```suggestion
```

File: samples/posix/uname/README.rst [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    uart:~$ uname -r
+    3.5.99
+    uart:~$ uname -v
+    zephyr-v3.5.0-3515-g10156f5f1d9c Jan  9 2024 22:23:04
+    uart:~$ uname -m
+    x86
+    uart:~$ uname -i
+    qemu_x86
+    uart:~$ uname -smi
+    Zephyr x86 qemu_x86
```
Comment (kartben) [inline samples/posix/uname/README.rst:56]: I don't think showing all the possible ways to run the sample is really necessary.
It certainly makes sense to have at least one with the multiple options though

```suggestion
    uart:~$ uname -a
    Zephyr zephyr 3.5.99 zephyr-v3.5.0-3515-g10156f5f1d9c Jan  9 2024 22:23:04 x86 qemu_x86
    uart:~$ uname -smi
    Zephyr x86 qemu_x86
```

**Review by kartben — CHANGES_REQUESTED** [review]
Thanks!
Additional note: line wrapping is a bit inconsistent in the two proposed files. I realize this might be due to copy-pasting from the other POSIX sample README, but it would make sense to adjust to have consistent 100-character long lines.

File: samples/posix/eventfd/README.rst (hunk 4 above)
Comment (cfriedt) [reply samples/posix/eventfd/README.rst:9]: `eventfd()` is not yet part of POSIX. It just happens to be lumped in with POSIX in Zephyr because it's available on most POSIX systems, is mostly dependent already on existing POSIX APIs, etc.

File: samples/posix/uname/README.rst (hunk 3 above)
Comment (cfriedt) [reply samples/posix/uname/README.rst:9]: @kartben - still on the backburner, but I think we will eventually add doxygen for all posix api calls, with links to the upstream specs. It's just such a big endeavour, that I would probably want to do it in an automated way.

**Comment by dougwfoster** [comment]
> LGTM - a great follow up commit or PR would be to list POSIX samples in the os services / portability / POSIX doc area.
> 
> Would be happy to re-approve if you wish to add that here, @dougwfoster ?

That's a good idea. I can add a **POSIX Sample** section there or list them in the **POSIX Applications in Zephyr** section.

File: samples/posix/uname/README.rst [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+This project outputs to the console.  It can be built and executed
+on QEMU as follows:
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/posix/uname
+   :host-os: unix
+   :board: qemu_x86
+   :goals: run
+   :compact:
+
```
[the remark this answers is not in the document]
Comment (dougwfoster) [reply samples/posix/uname/README.rst:27]: I noticed the uname sample does not have a Makefile.host file. Is there a specific reason it doesn't have one? I will add one and see if it builds.

File: samples/posix/eventfd/README.rst [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+   :host-os: unix
+   :board: qemu_x86
+   :goals: run
+   :compact:
+
+For comparison, to build directly for your host OS if it is POSIX compliant (for ex. Linux):
+
+.. code-block:: console
+
+   make -f Makefile.host
```
[the remark this answers is not in the document]
Comment (dougwfoster) [reply samples/posix/eventfd/README.rst:32]: Thanks for pointing bringing this to my attention. It looks like the Makefile.host file is actually set up to work only if the cwd is **samples/posix/eventfd**, so the command to build it would be `make -f Makefile.host`. I can update the Makefile.host files so it can be built with  `make -f samples/posix/eventfd/Makefile.host`, similar to how the west build would work, and update the instructions.

**Comment by cfriedt** [comment]
@dougwfoster - I think the samples portion should probably be automatically built (as long as the .rst files are captured by the main sample rst)

From the os services / portability / POSIX page, maybe just a link back to the samples would be good?

@kartben could probably be more specific.

File: samples/posix/uname/README.rst (hunk 3 above)
Comment (kartben) [reply samples/posix/uname/README.rst]: @cfriedt you're suggesting writing original docs, right? Because I don't think the Open Group license would allow us to reuse/redistribute their documentation?

File: samples/posix/eventfd/README.rst [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+   :goals: run
+   :compact:
+
+For comparison, to build directly for your host OS if it is POSIX compliant (for ex. Linux):
+
+.. code-block:: console
+
+   make -f samples/posix/eventfd/Makefile.host
+
+The make output file will be located in samples/posix/eventfd.
```
Comment (kartben) [inline samples/posix/eventfd/README.rst:31]: non blocking: kind of not ideal that this adds a new untracked file to the git repo, but I guess it was already like that before.

**Review by kartben — DISMISSED** [review]

File: samples/posix/eventfd/README.rst [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+********
+
+This sample application demonstrates using the POSIX eventfd() function to create a file descriptor,
+which can be used for event notification. The returned file descriptor is used with write/read calls
+and write/read values are output to the console.
+
+Building and Running
+********************
+
+This project outputs to the console.  It can be built and executed on QEMU as follows:
```
Comment (ycsin) [inline samples/posix/eventfd/README.rst:16]:
```suggestion
This project outputs to the console. It can be built and executed on QEMU as follows:
```

File: samples/posix/uname/README.rst [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+********
+
+In this sample application, the POSIX `uname()`_ function is used to acquire system information and
+it is output to the console. Additionally, uname is added as a shell command and system information
+is displayed according to the option(s) provided for the command.
+
+Building and Running
+********************
+
+This project outputs to the console.  It can be built and executed on QEMU as follows:
```
Comment (ycsin) [inline samples/posix/uname/README.rst:16]:
```suggestion
This project outputs to the console. It can be built and executed on QEMU as follows:
```

**Review by ycsin — APPROVED** [review]

File: samples/posix/eventfd/Makefile.host [hunk 13]
```diff
@@ -1,4 +1,4 @@
 # This makefile builds the sample for a POSIX system, like Linux
 
-eventfd: src/main.c
-	$(CC) $^ -o $@
+eventfd: samples/posix/eventfd/src/main.c
```
Comment (ycsin) [inline samples/posix/eventfd/Makefile.host:3]: > Updated Makefile.hostfile for samples to be used with make from 'zephyr' directory, similar to west build command.

I find this change somewhat unnecessary, it can't be compared with the `west build` command because `west` works anywhere (either in the `zephyr` folder or in the `zephyr/samples/posix/eventfd` folder). Changing the path in this Makefile doesn't fix its limitation, it's just a matter of preference (i.e. I usually build in the `zephyrproject` folder, this change wouldn't benefit me) and it deviates from other existing `Makefile.host`s in the `net/socket` examples.

Also, before this change, a developer can copy the sample to anywhere and `make` would work as-is, after this change, the sample must be in the relative path for the Makefile to work.

TL;DR: This change only increases the limitation of the Makefile

**Review by ycsin — CHANGES_REQUESTED** [review]

File: samples/posix/gettimeofday/Makefile.host [hunk 14]
```diff
@@ -1,4 +1,4 @@
 # This makefile builds the sample for a POSIX system, like Linux
 
-gettimeofday: src/main.c
-	$(CC) $^ -o $@
+gettimeofday: samples/posix/gettimeofday/src/main.c
+	$(CC) $^ -o samples/posix/gettimeofday/$@
```
Comment (ycsin) [inline samples/posix/gettimeofday/Makefile.host:4]: If we output the binary to a `build` folder, then it will be ignored by git and won't be an untracked file cc @kartben :

```suggestion
gettimeofday: src/main.c
	mkdir -p build
	$(CC) $^ -o build/$@
```

Comment (dougwfoster) [reply samples/posix/gettimeofday/Makefile.host]: Thanks for the suggestion. I made this change in my latest update.

File: samples/posix/uname/README.rst (hunk 8 above)
[the remark this answers is not in the document]
Comment (ycsin) [reply samples/posix/uname/README.rst:24]: @dougwfoster this now can be done since #67648 is merged

**Review by cfriedt — APPROVED** [review]
LGTM - the only slight inaccuracy is that events is not POSIX (yet), even though it is available on many POSIX operating systems.

Otherwise, I think maybe @kartben could potentially offer some style suggestions?

**Review by ycsin — APPROVED** [review]

**Review by aescolar — APPROVED** [review]
